src/ir/stack-utils.h

```diff
@@ -33,7 +33,10 @@
 //     stack type of each instruction. Pops may not have `unreachable` type.
 //
 //  4. Only control flow structures and instructions that have polymorphic
-//     unreachable behavior in WebAssembly may have unreachable type.
+//     unreachable behavior in WebAssembly may have unreachable type. Blocks may
+//     be unreachable when they are not branch targets and when they have an
+//     unreachable child. Note that this means a block may be unreachable even
+//     if it would otherwise have a concrete type, unlike in Binaryen IR.
```

Comment:
Where do we do this? Poppy IR doesn't seem to have its own `Block::finalize()` or something and it shares the parser with Binaryen IR, aren't block types in Poppy IR assigned in the same way as that of Binaryen IR?

Reply:
Eventually we will need to add alternate logic to `Block::finalize()` to handle Poppy IR. The reason I haven't done this yet is because the parser only calls `Block::finalize(Type)`, so none of the tests call `Block::finalize()`, so we would have no test coverage. The next PR will add the stackification pass, which will call refinalize, so I will implement this with test coverage then.

Reply:
For example,
```wat
(block (result i32)
  (unreachable)
)
```
According to your explanation, this block's type will be `i32` in Binaryen IR and `unreachable` in Poppy IR. This block has a result type so `Block::finalize(type)` will be called. But the current `Block::finalize(type)` [respects](https://github.com/WebAssembly/binaryen/blob/41ea543093ed734c73e5202c58c899be447b3223/src/wasm/wasm.cpp#L366) the `type` given as the argument and does not check instructions inside for unreachables, no?

What I meant was, as you said, given that we only use `Block::finalize(type)` for now, how we determine the unreachability of a block will be the same for Binaryen IR and Poppy IR, so this comments block does not seem to match the current implementation. I think I might be missing something.

Reply:
In both Binaryen IR and Poppy IR, the block in your example can have either `i32` or `unreachable` as its type. In Binaryen IR, the parser will give it `i32` type and subsequent optimizations will change it to have `unreachable` type. In Poppy IR, the parser will similarly give it `i32` type, but we don't yet have any code that will change the type to `unreachable` (even though that would be allowed).

The difference between Poppy IR and Binaryen IR is in a block like this:

```
(block
  (unreachable)
  (i32.const 0)
)
```

In Binaryen IR, it would be illegal for that block to have `unreachable` type, but in Poppy IR it could still have either `unreachable` or `i32` type.

Reply:
I see, thanks. Is there any particular reason for this distinction?

Reply:
Yes, it has to do with type inference in Poppy IR, which we don't do in Binaryen IR. Consider the following function:

```
(func (result i32)
  (block $body
    (unreachable)
    (i32.const 0)
  )
)
```

Binaryen IR unconditionally gives `unreachable` blocks `none` type when writing them out, so if `$body` were allowed to be `unreachable` in Binaryen IR, there resulting binary would be invalid. In contrast, Poppy IR does type inference before emitting binaries, which correctly gives `$body` an `i32` type.